### PR TITLE
add infactory sensor

### PR DESCRIPTION
**Pull request by siriuzwhite**
Related to https://github.com/merbanan/rtl_433/issues/591

### CODE REVIEW

File: src/devices/infactory.c [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ * Outdoor sensor transmits data temperature, humidity.
+ * Transmissions also includes an id. The sensor transmits
+ * every 60 seconds 6 packets.
+ *
+ * 0000 1111 | 0011 0000 | 0101 1100 | 1110 0111 | 0110 0001
+ * xxxx xxxx | cccc cccc | tttt tttt | tttt hhhh | hhhh ????
+ *
+ * x - ID // changes on battery switch
+ * c - Unknown Checksum (changes on every transmit if the other values are different)
+ * h - Humidity // each nibble represents one diggit of percentage value
```
Comment (zuckschwerdt) [inline src/devices/infactory.c]: BCD-encoded, each nibble is one digit

File: src/devices/infactory.c [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * Usage:
+ *
+ *    # rtl_433 -f 434052000 -R 91 -F json:log.json
+ *
+ */
+
+#include "rtl_433.h"
+#include "util.h"
+#include "pulse_demod.h"
```
Comment (zuckschwerdt) [inline src/devices/infactory.c]: you don't need pulse_demod header

File: src/devices/infactory.c [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+#include "data.h"
+
+static int infactory_callback(bitbuffer_t *bitbuffer) {
+
+    bitrow_t *bb = bitbuffer->bb;
+    uint8_t *b = bb[0];
+    data_t *data;
+
+    char time_str[LOCAL_TIME_BUFLEN];
+    local_time_str(0, time_str);
```
Comment (zuckschwerdt) [inline src/devices/infactory.c]: move this down just before data_make

File: src/devices/infactory.c [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    if (bitbuffer->bits_per_row[0] == 40) {
```
Comment (zuckschwerdt) [inline src/devices/infactory.c]: make this ... != 40) {
return 0;
}

File: src/devices/infactory.c [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      uint hum1 = b[3] & 0xF;
+      uint hum2 = b[4] >> 4;
+      uint humidity = hum1 * 10 + hum2;
+
+      uint8_t temp1 = b[2];
+      uint temp2 = b[3] >> 4;
+      uint16_t temp = (temp1 << 4) | temp2;
+
+      double temp_f = (double)temp / 10 - 90;
+      double temp_c = (temp_f - 32)*5/9;
```
Comment (zuckschwerdt) [inline src/devices/infactory.c]: do not convert, output deg F

File: src/devices/infactory.c [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    bitrow_t *bb = bitbuffer->bb;
+    uint8_t *b = bb[0];
+    data_t *data;
+
+    char time_str[LOCAL_TIME_BUFLEN];
+    local_time_str(0, time_str);
+
+    if (bitbuffer->bits_per_row[0] == 40) {
+
+      //bitbuffer_print(bitbuffer);
```
Comment (zuckschwerdt) [inline src/devices/infactory.c]: remove your debug notes

File: src/devices/infactory.c [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+      uint16_t temp = (temp1 << 4) | temp2;
+
+      double temp_f = (double)temp / 10 - 90;
+      double temp_c = (temp_f - 32)*5/9;
+
+
+      data = data_make( "time",		"",				DATA_STRING,	time_str,
+          "model",         "",	   DATA_STRING, "inFactory sensor",
+          "id",      "ID",   DATA_FORMAT, "%u", DATA_INT, id,
+          "temperature_c", "Temperature",DATA_FORMAT, "%.02f °C", DATA_DOUBLE, temp_c,
```
Comment (zuckschwerdt) [inline src/devices/infactory.c]: temperature_F (capital "F")

File: src/devices/infactory.c [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+    return 1;
+}
+
+
+static char *output_fields[] = {
+    "time",
+    "model"
+    "id",
+    "temperature_c",
```
Comment (zuckschwerdt) [inline src/devices/infactory.c]: temperature_F

File: src/devices/infactory.c [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+};
+
+r_device infactory = {
+    .name          = "inFactory",
+    .modulation    = OOK_PULSE_PPM_RAW,
+    .short_limit   = 3000, // Threshold between short and long gap [us]
+    .long_limit    = 5000, //  Maximum gap size before new row of bits [us]
+    .reset_limit   = 6000, // Maximum gap size before End Of Message [us].
+    .json_callback = &infactory_callback,
+    .disabled      = 0,
```
Comment (zuckschwerdt) [inline src/devices/infactory.c]: set disabled to 1 until the checksum is figured out.

**Comment by zuckschwerdt** [comment]
It's custom style for devices to declare variables at the top.
And you don't need intermediate vars and type conversions for temp & hum. E.g.
```
humidity = (b[3] & 0x0F) * 10 + (b[4] >> 4); // BCD
temp = (b[2] << 4) | (b[3] >> 4);
temp_f = (double)temp / 10 - 90;
```
will do and is still very readable.

Looks good to me otherwise.

**Comment by siriuzwhite** [comment]
Thanks for you help :-)

File: src/devices/infactory.c [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    double temp_f;
+
+    if (bitbuffer->bits_per_row[0] != 40) {
+      return 0;
+    }
+
+    id = b[0];
+    humidity = (b[3] & 0x0F) * 10 + (b[4] >> 4); // BCD
+    temp = (b[2] << 4) | (b[3] >> 4);
+    temp_f = (double)temp / 10 - 90;
```
Comment (merbanan) [inline src/devices/infactory.c]: Use (float)

File: src/devices/infactory.c [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+    bitrow_t *bb = bitbuffer->bb;
+    uint8_t *b = bb[0];
+    data_t *data;
+
+    char time_str[LOCAL_TIME_BUFLEN];
+    uint8_t id;
+    uint humidity;
+    uint16_t temp;
+    double temp_f;
```
Comment (merbanan) [inline src/devices/infactory.c]: Use float

File: src/devices/infactory.c [hunk 12]
```diff
@@ -0,0 +1,79 @@
+/*
```
Comment (merbanan) [inline src/devices/infactory.c]: Add a license and copyright header.

**Comment by merbanan** [comment]
Looks good. Just fix the comments and I will merge the code.

**Comment by zuckschwerdt** [comment]
The checksum seems to be a CRC-4 with polynomial 0x3 and initial 0x1. Maybe we can roll that into this PR and switch the disabled to 0 :) I'll run a few tests on the CRC and get back to you.

**Comment by zuckschwerdt** [comment]
Actually don't wait for me to figure the CRC out. It will take a few days to sort this properly and we can merge it later on.

**Comment by siriuzwhite** [comment]
Still produces 6 identical entries for each single transmission. Is this "ok" for disabled = 0 ?

**Comment by zuckschwerdt** [comment]
`disabled = 1` is needed because there are no integrity checks, there would be a lot of false positives. Remember to use -R 91 when using this device.
Usually a device would recieve the (6 here) packets in multiple rows. But the timing here is too slow to allow that, i.e. the packet distance is too big.

**Pull request merged**